File: hooks/_lib/pkg_rewrite.py
#!/usr/bin/env python3
"""Package manager transparent correction: npm/yarn→pnpm, pip→uv.

Reads a shell command from stdin, prints the corrected command to stdout.
Prints empty string if no correction is needed.

Only rewrites simple single commands — chain commands (&&, ||, ;, pipes)
are not corrected to avoid mistakenly modifying complex pipelines.
"""
import sys
import re

cmd = sys.stdin.read().strip()
corrected = None

# Skip complex commands
if not re.search(r"&&|&|\|\||;|[|<>\n\r]|\$\(|`", cmd):

    # npm install (no parameters) → pnpm install
    if re.match(r"^npm\s+(?:install|i)\s*$", cmd):
        corrected = "pnpm install"

    # npm install/add <packages>
    elif re.match(r"^npm\s+(?:install|i|add)\s+", cmd):
        rest = re.sub(r"^npm\s+(?:install|i|add)\s+", "", cmd).strip()
        tokens = rest.split()

        KNOWN_FLAGS = {"--save-dev", "-D", "--save", "-S", "--save-optional", "-O", "--save-exact", "-E"}

        is_global = any(t in ("-g", "--global") or t.startswith("--location=global") for t in tokens)
        unknown_flags = [t for t in tokens if t.startswith("-") and t not in KNOWN_FLAGS]
        packages = [t for t in tokens if not t.startswith("-")]

        if packages and not is_global and not unknown_flags:
            pnpm_flags = []
            for t in tokens:
                if t in ("--save-dev", "-D"):
                    pnpm_flags.append("-D")
                elif t in ("--save-optional", "-O"):
                    pnpm_flags.append("-O")
                elif t in ("--save-exact", "-E"):
                    pnpm_flags.append("--save-exact")
            corrected = "pnpm add " + " ".join(pnpm_flags + packages).strip()

    # yarn install (no parameters) → pnpm install
    elif re.match(r"^yarn\s+install\s*$", cmd):
        corrected = "pnpm install"

    # yarn add <packages> → pnpm add <packages>
    elif re.match(r"^yarn\s+add\s+", cmd):
        rest = re.sub(r"^yarn\s+add\s+", "", cmd)
        tokens = rest.split()
        YARN_KNOWN_FLAGS = {"-D", "--dev", "--save-dev", "-O", "--optional",
                            "-E", "--exact", "-P", "--save-peer",
                            "-W", "--ignore-workspace-root-check"}
        packages = [t for t in tokens if not t.startswith("-")]
        unknown_flags = [t for t in tokens if t.startswith("-") and t not in YARN_KNOWN_FLAGS]
        if packages and not unknown_flags:
            pnpm_flags = []
            for t in tokens:
                if t in ("-D", "--dev", "--save-dev"):
                    pnpm_flags.append("-D")
                elif t in ("-O", "--optional"):
                    pnpm_flags.append("-O")
                elif t in ("-E", "--exact"):
                    pnpm_flags.append("--save-exact")
                elif t in ("-P", "--save-peer"):
                    pnpm_flags.append("--save-peer")
                elif t in ("-W", "--ignore-workspace-root-check"):
                    pnpm_flags.append("-w")
            corrected = "pnpm add " + " ".join(pnpm_flags + packages).strip()

    # pip install / pip3 install → uv pip install
    elif re.match(r"^pip3?\s+install\s+", cmd):
        rest = re.sub(r"^pip3?\s+install\s+", "", cmd)
        tokens = rest.split()
        PIP_KNOWN_FLAGS = {"-r", "--requirement", "-e", "--editable",
                           "-U", "--upgrade", "--pre", "--no-deps",
                           "-i", "--index-url", "--extra-index-url", "--no-index",
                           "-f", "--find-links", "-c", "--constraint",
                           "-v", "--verbose", "-q", "--quiet", "-t", "--target"}
        unknown_flags = [t for t in tokens if t.startswith("-") and t not in PIP_KNOWN_FLAGS]
        if not unknown_flags:
            corrected = "uv pip install " + rest

    # python -m pip install / python3 -m pip install → uv pip install
    elif re.match(r"^python3?\s+-m\s+pip\s+install\s+", cmd):
        rest = re.sub(r"^python3?\s+-m\s+pip\s+install\s+", "", cmd)
        tokens = rest.split()
        PIP_KNOWN_FLAGS = {"-r", "--requirement", "-e", "--editable",
                           "-U", "--upgrade", "--pre", "--no-deps",
                           "-i", "--index-url", "--extra-index-url", "--no-index",
                           "-f", "--find-links", "-c", "--constraint",
                           "-v", "--verbose", "-q", "--quiet", "-t", "--target"}
        unknown_flags = [t for t in tokens if t.startswith("-") and t not in PIP_KNOWN_FLAGS]
        if not unknown_flags:
            corrected = "uv pip install " + rest

print(corrected or "")

File: vg-helper/src/pkg_rewrite.rs
//! Package manager transparent correction: npm/yarn→pnpm, pip→uv.
//! Canonical implementation. hooks/_lib/pkg_rewrite.py is a deprecated fallback.

use regex::Regex;
use std::io::{self, Read};

type Result = std::result::Result<(), Box<dyn std::error::Error>>;

/// Reads a shell command from stdin, prints corrected command or empty string.
pub fn run(_args: &[String]) -> Result {
    let mut cmd = String::new();
    io::stdin().read_to_string(&mut cmd)?;
    let cmd = cmd.trim();

    println!("{}", rewrite_command(cmd).unwrap_or_default());
    Ok(())
}

fn rewrite_command(cmd: &str) -> Option<String> {
    // Skip complex commands (&&, ||, ;, pipe, redirection, $(), backtick)
    let complex = Regex::new(r"&&|&|\|\||;|[|<>\n\r]|\$\(|`").ok()?;
    if complex.is_match(cmd) {
        return None;
    }

    try_npm(cmd)
        .or_else(|| try_yarn(cmd))
        .or_else(|| try_pip(cmd))
        .or_else(|| try_python_pip(cmd))
}

fn try_npm(cmd: &str) -> Option<String> {
    let bare_install = Regex::new(r"^npm\s+(?:install|i)\s*$").ok()?;
    if bare_install.is_match(cmd) {
        return Some("pnpm install".into());
    }

    let pkg_install = Regex::new(r"^npm\s+(?:install|i|add)\s+").ok()?;
    if !pkg_install.is_match(cmd) {
        return None;
    }
    let rest = pkg_install.replace(cmd, "").trim().to_string();
    let tokens: Vec<&str> = rest.split_whitespace().collect();

    const KNOWN: &[&str] = &[
        "--save-dev",
        "-D",
        "--save",
        "-S",
        "--save-optional",
        "-O",
        "--save-exact",
        "-E",
    ];
    if tokens
        .iter()
        .any(|t| *t == "-g" || *t == "--global" || t.starts_with("--location=global"))
    {
        return None;
    }
    if tokens
        .iter()
        .any(|t| t.starts_with('-') && !KNOWN.contains(t))
    {
        return None;
    }
    let packages: Vec<&str> = tokens
        .iter()
        .filter(|t| !t.starts_with('-'))
        .copied()
        .collect();
    if packages.is_empty() {
        return None;
    }

    let mut flags = Vec::new();
    for t in &tokens {
        match *t {
            "--save-dev" | "-D" => flags.push("-D"),
            "--save-optional" | "-O" => flags.push("-O"),
            "--save-exact" | "-E" => flags.push("--save-exact"),
            _ => {}
        }
    }
    flags.extend(packages);
    Some(format!("pnpm add {}", flags.join(" ")))
}

fn try_yarn(cmd: &str) -> Option<String> {
    let bare = Regex::new(r"^yarn\s+install\s*$").ok()?;
    if bare.is_match(cmd) {
        return Some("pnpm install".into());
    }

    let add = Regex::new(r"^yarn\s+add\s+").ok()?;
    if !add.is_match(cmd) {
        return None;
    }
    let rest = add.replace(cmd, "").to_string();
    let tokens: Vec<&str> = rest.split_whitespace().collect();

    const KNOWN: &[&str] = &[
        "-D",
        "--dev",
        "--save-dev",
        "-O",
        "--optional",
        "-E",
        "--exact",
        "-P",
        "--save-peer",
        "-W",
        "--ignore-workspace-root-check",
    ];
    if tokens
        .iter()
        .any(|t| t.starts_with('-') && !KNOWN.contains(t))
    {
        return None;
    }
    let packages: Vec<&str> = tokens
        .iter()
        .filter(|t| !t.starts_with('-'))
        .copied()
        .collect();
    if packages.is_empty() {
        return None;
    }

    let mut flags = Vec::new();
    for t in &tokens {
        match *t {
            "-D" | "--dev" | "--save-dev" => flags.push("-D"),
            "-O" | "--optional" => flags.push("-O"),
            "-E" | "--exact" => flags.push("--save-exact"),
            "-P" | "--save-peer" => flags.push("--save-peer"),
            "-W" | "--ignore-workspace-root-check" => flags.push("-w"),
            _ => {}
        }
    }
    flags.extend(packages);
    Some(format!("pnpm add {}", flags.join(" ")))
}

fn try_pip(cmd: &str) -> Option<String> {
    let re = Regex::new(r"^pip3?\s+install\s+").ok()?;
    if !re.is_match(cmd) {
        return None;
    }
    let rest = re.replace(cmd, "").to_string();
    check_pip_flags(&rest).map(|_| format!("uv pip install {rest}"))
}

fn try_python_pip(cmd: &str) -> Option<String> {
    let re = Regex::new(r"^python3?\s+-m\s+pip\s+install\s+").ok()?;
    if !re.is_match(cmd) {
        return None;
    }
    let rest = re.replace(cmd, "").to_string();
    check_pip_flags(&rest).map(|_| format!("uv pip install {rest}"))
}

fn check_pip_flags(rest: &str) -> Option<()> {
    const KNOWN: &[&str] = &[
        "-r",
        "--requirement",
        "-e",
        "--editable",
        "-U",
        "--upgrade",
        "--pre",
        "--no-deps",
        "-i",
        "--index-url",
        "--extra-index-url",
        "--no-index",
        "-f",
        "--find-links",
        "-c",
        "--constraint",
        "-v",
        "--verbose",
        "-q",
        "--quiet",
        "-t",
        "--target",
    ];
    let tokens: Vec<&str> = rest.split_whitespace().collect();
    if tokens
        .iter()
        .any(|t| t.starts_with('-') && !KNOWN.contains(t))
    {
        return None;
    }
    Some(())
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn npm_bare_install_rewrites_to_pnpm_install() {
        assert_eq!(rewrite_command("npm install"), Some("pnpm install".into()));
        assert_eq!(rewrite_command("npm i"), Some("pnpm install".into()));
    }

    #[test]
    fn npm_package_install_rewrites_supported_flags() {
        assert_eq!(
            rewrite_command("npm install --save-dev --save-exact vitest @types/node"),
            Some("pnpm add -D --save-exact vitest @types/node".into())
        );
        assert_eq!(
            rewrite_command("npm add -O sharp"),
            Some("pnpm add -O sharp".into())
        );
    }

    #[test]
    fn npm_global_or_unknown_flags_do_not_rewrite() {
        assert_eq!(rewrite_command("npm install -g typescript"), None);
        assert_eq!(
            rewrite_command("npm install --legacy-peer-deps react"),
            None
        );
    }

    #[test]
    fn yarn_install_and_add_rewrite_to_pnpm() {
        assert_eq!(rewrite_command("yarn install"), Some("pnpm install".into()));
        assert_eq!(
            rewrite_command("yarn add --dev --exact eslint"),
            Some("pnpm add -D --save-exact eslint".into())
        );
        assert_eq!(
            rewrite_command("yarn add -P -W react"),
            Some("pnpm add --save-peer -w react".into())
        );
    }

    #[test]
    fn yarn_unknown_flags_do_not_rewrite() {
        assert_eq!(rewrite_command("yarn add --frozen-lockfile react"), None);
    }

    #[test]
    fn pip_install_rewrites_known_forms() {
        assert_eq!(
            rewrite_command("pip install requests"),
            Some("uv pip install requests".into())
        );
        assert_eq!(
            rewrite_command("pip3 install -r requirements.txt"),
            Some("uv pip install -r requirements.txt".into())
        );
    }

    #[test]
    fn python_module_pip_rewrites_known_forms() {
        assert_eq!(
            rewrite_command("python -m pip install -e ."),
            Some("uv pip install -e .".into())
        );
        assert_eq!(
            rewrite_command("python3 -m pip install --upgrade build"),
            Some("uv pip install --upgrade build".into())
        );
    }

    #[test]
    fn pip_unknown_flags_do_not_rewrite() {
        assert_eq!(rewrite_command("pip install --user requests"), None);
        assert_eq!(
            rewrite_command("python3 -m pip install --user requests"),
            None
        );
    }

    #[test]
    fn complex_commands_do_not_rewrite() {
        for cmd in [
            "npm install && npm test",
            "pip install requests | cat",
            "yarn add react > out.txt",
            "npm install $(cat package.txt)",
            "npm install `cat package.txt`",
        ] {
            assert_eq!(rewrite_command(cmd), None, "{cmd}");
        }
    }
}
